bundles/org.openhab.binding.solax/src/main/java/org/openhab/binding/solax/internal/SolaxLocalAccessHandler.java

```diff
@@ -89,18 +93,26 @@ public void initialize() {
 
     private void retrieveData() {
         try {
-            String rawJsonData = localHttpConnector.retrieveData();
-            logger.debug("Raw data retrieved = {}", rawJsonData);
-
-            if (rawJsonData != null && !rawJsonData.isEmpty()) {
-                updateFromData(rawJsonData);
+            if (retrieveDataCallLock.tryLock()) {
+                String rawJsonData = localHttpConnector.retrieveData();
+                logger.debug("Raw data retrieved = {}", rawJsonData);
+
+                if (rawJsonData != null && !rawJsonData.isEmpty()) {
+                    updateFromData(rawJsonData);
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                            SolaxBindingConstants.I18N_KEY_OFFLINE_COMMUNICATION_ERROR_JSON_CANNOT_BE_RETRIEVED);
+                }
             } else {
-                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
-                        SolaxBindingConstants.I18N_KEY_OFFLINE_COMMUNICATION_ERROR_JSON_CANNOT_BE_RETRIEVED);
+                logger.debug("Unable to retrieve data because a request is already in progress.");
             }
         } catch (IOException e) {
             logger.debug("Exception received while attempting to retrieve data via HTTP", e);
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        } finally {
+            if (retrieveDataCallLock.isHeldByCurrentThread()) {
```

Comment:
To avoid the `isHeldByCurrentThread`, I would imagine rather this code:
```
        if (retrieveDataCallLock.tryLock()) {
           try {
                String rawJsonData = localHttpConnector.retrieveData();
                logger.debug("Raw data retrieved = {}", rawJsonData);

                if (rawJsonData != null && !rawJsonData.isEmpty()) {
                    updateFromData(rawJsonData);
                } else {
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
SolaxBindingConstants.I18N_KEY_OFFLINE_COMMUNICATION_ERROR_JSON_CANNOT_BE_RETRIEVED);
                }
            } catch (IOException e) {
                logger.debug("Exception received while attempting to retrieve data via HTTP", e);
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
            } finally {
                retrieveDataCallLock.unlock();
            }
        } else {
            logger.debug("Unable to retrieve data because a request is already in progress.");
        }
```

WDYT ?

Reply:
Agree. Thanks for the suggestion... I'll refactor it today.

Reply:
Not sure why I did it like this initially, looking at what you suggested, actually... :) Looks way better.